Added nodename rule to custom source

slintes opened the pull request:
# What
With this PR we add a new `nodename` rule for the `custom` source for the NFD worker, which will apply features based on  nodes' hostnames.

For making more dynamic configuration by potentially multiple parties easier, `custom` source rules can also be read from the `/etc/kubernetes/node-feature-discovery/custom.d` directory now (probabaly mostly populated by mounted ConfigMaps).

Also an optional `value` field was added, which allows overruling the default `true` value for matching features if set.

# Details

## Example
Consider this config:
```
- name: "feature.by.nodename"
  # default value is true
  matchOn:
    - nodename:
      - worker-0
      - worker-1
- name: "another.features.by.nodename.regex"
  value: customValue
  matchOn:
    - nodename:
      - wor.*-0
```

The config will apply the `feature.by.nodename=true` feature to the `worker-0` and `worker-1` nodes, and the `another.features.by.nodename.regex=customValue` feature to all nodes matching the `wor.*-0` regex.

## Implementation
- the mounting of additional ConfigMaps into the NFD worker pod is a bit cumbersome, but can be automated with the NFD operator in a follow up (e.g. mount all ConfigMaps in the NFD namespace which match a certain name pattern).

# Usecase
A customer with many nodes is able to set the nodename according the node's purpose during node deployment, but not to add the needed labels for workload scheduling. The nodename is the only metadata available for determining the node's purpose.
With this new source the customer can put the desired nodename - label mapping(s) into a ConfigMap, and let NFD do the work of putting the labels on the nodes.

/cc @zvonkok @ArangoGutierrez @marquiz

ArangoGutierrez commented:
/ok-to-test

ArangoGutierrez commented:
/assign

marquiz commented:
Thanks @slintes for the patchset!

I see the use case. What I immediately started to think is that would it be possible to extend the existing `custom` source, instead of adding a new one(?) Is there need for multiple configmaps or could the single nfd-worker configmap instance be used instead? If multiple configs would be required we could enable a config dir something like `/etc/kubernetes/node-feature-discovery/custom.d` (similar to `source.d` and `features.d`, currently). Currently, the limitation with the custom feature rules is that there can only be one label per rule but I think that could be extended with a `labels` field (similar to what you have in this PR). WDYT?

ArangoGutierrez reviewed (changes requested):
Patch is looking good, I need to run a couple of tests before digging deeper.
At a first glance, this patch requires unit testing for the big chunks of GO code being added, could you please include at least some minor unit tests for this new features, thanks

marquiz commented:
> Patch is looking good, I need to run a couple of tests before digging deeper.
> At a first glance, this patch requires unit testing for the big chunks of GO code being added, could you please include at least some minor unit tests for this new features, thanks

I really want to explore extending the existing source(s) (`custom`) before going forward with this as is

slintes commented:
@marquiz Thanks for the feedback, and I understand it is preferable to extend an existing source.

The hostname - label matching rules are potentially very dynamic, and there might even be multiple parties which want to add or modify the rules (the exact requirements are vague unfortunately, that's why I went with a very flexible implementation).
Having to modify the main worker config for this would need some careful coordination. Having dedicated ConfigMaps for each party simplifies configuration from the user's POV a lot in this scenario IMHO.

For having multiple ConfigMaps you proposed a `/etc/kubernetes/node-feature-discovery/custom.d` dir... do I understand correctly that
- those additional ConfigMaps should be mounted into that directory
- the custom source will use them additionally to the existing config (similar to how it's implemented in the my current approach)
- but in contrast to the current impl those ConfigMaps will contain the rules in the same format like in the existing worker config?

I think multiple labels per rule are not really needed... it was just easy to implement. But I don't see an actual usecase. If you actually need multiple labels, it will be very few only, which can still be configured with multiple rules. Not worth a modification of the current config structure IMHO.

@ArangoGutierrez yes, I will add unit tests as soon as we have an agreement to the way forward. I expected the e2e test to run on the PR as well, that's why I focused on that one in the 1st iteration.

marquiz commented:
> The hostname - label matching rules are potentially very dynamic, and there might even be multiple parties which want to add or modify the rules (the exact requirements are vague unfortunately, that's why I went with a very flexible implementation).

Yes, I understand this, and makes sense. That's why I proposed to extend the custom source to support multiple config files.

> For having multiple ConfigMaps you proposed a `/etc/kubernetes/node-feature-discovery/custom.d` dir... do I understand correctly that
> 
> * those additional ConfigMaps should be mounted into that directory

Yes, that's right.

> * the custom source will use them additionally to the existing config (similar to how it's implemented in the my current approach)

Yes, rules could be defined in the nfd-worker's main config file like we already have. In addition to that, additional rules could be specified by adding rule files under the `custom.d`.

> * but in contrast to the current impl those ConfigMaps will contain the rules in the same format like in the existing worker config?

Yes, the format would be similar. That would make it possible to define any rules there and for example combine hostname/nodename matching with other rules. E.g.
```yaml
- name: "my.host.feature"
  matchOn:
    - hostName: ["host-1", "host-2"]
    - nodeName: ["host-1", "host-2"]
- name: "my.kernel.feature"
  matchOn:
    - loadedKMod: ["kmod-1"]
- name: "my.combo.feature"
  matchOn:
    - hostName: ["host-1", "host-2"]
    - pciId:
        vendor: ["8086"]
        device: ["1234"]
```
I think the `hostName`/`nodeName` could be a treated as a regexp to make matching super flexible.

> I think multiple labels per rule are not really needed... it was just easy to implement.

OK, even better. And, I think that could be added later, then, if there's a need for that. We might want to add a `value` field for the custom config, though (IIRC, we don't we have that at the moment). All labels just get value `true`.

slintes commented:
> I think the hostName/nodeName could be a treated as a regexp to make matching super flexible.

Yes, that's a requirement actually :)

> We might want to add a value field ... All labels just get value true

oh, good point! Didn't notice that yet.

slintes commented:
- refactored to custom source
- added possibility to set an optional label value, defaults to the old `true` value
- rebased
- unit test and doc update pending
- e2e test succeeds in a local test

slintes commented:
forgot this:

- added possibility to read more custom source configs from additional ConfigMaps
- updated PR description to reflect the changes

:)

marquiz commented:
Really cool, thanks @slintes ! I'll take a detailed look as soon as I have time. Trying to do that on Friday but I'm afraid tomorrow is lost to other things

adrianchiris commented on source/custom/custom.go:
```diff
@@ -141,6 +148,16 @@ func (s Source) discoverFeature(feature FeatureSpec) (bool, error) {
 				continue
 			}
 		}
+		// hostname rule
```
not related to this commit, but we could possibly do something nifty with reflect package to avoid this repetitive code.

slintes replied:
I gave it a try. It works, but the code is a bit ugly IMHO. But that might be personal preference. You can see it at [1].
In this PR I went with a different approach for code de-deduplication, see last commit. It still needs one line with reflection, I didn't find another to filter out the nil values, but overall its better readable. With the downside of still needing to touch it for new rules though.

WDYT?

[1] https://github.com/slintes/node-feature-discovery/commit/ad705f9e14d71508d367ba7cc83a13e0e198f0dd#diff-1b283f9d332ba27ed8b09eb6642c7bf9c9d552d45f89d485c590c7c1619f6b7dR101

adrianchiris replied:
> I gave it a try. It works, but the code is a bit ugly IMHO

Doesnt seem too bad to me, but your current change is OK if you feel it is better.

marquiz replied:
I think the refactor in the last patch of this PR is good. I'm not a big fan of reflect but I guess there's no nice and easy way around it here.

slintes replied:
I'll keep the current version version then :slightly_smiling_face:

adrianchiris commented on source/custom/rules/hostname.go:
```diff
@@ -0,0 +1,54 @@
+
+var (
+	// TODO is it safe to assume nodeName == hostname??
+	nodeName = os.Getenv("NODE_NAME")
```
any reason not to use `/etc/hostname` ?

`NODE_NAME` is set as the k8s node object name, im sure there is some kubelet flag that allows to advertise the node under a different name somewhere.

from kubelet --help:
`--hostname-override string                                 If non-empty, will use this string as identification instead of the actual hostname.`

If these are indeed non synonymous maybe the Rule should be NodeNameRule ?

slintes replied:
Very good point, thanks! I need to clarify this.

slintes replied:
after clarifying I switch to nodename everywhere :)

slintes commented on nfd-daemonset-combined.yaml.template:
```diff
@@ -227,4 +227,10 @@ data:
     #            device: ["1014", "1017"]
     #          loadedKMod : ["vendor_kmod1", "vendor_kmod2"]
-### <NFD-WORKER-CONF-END-DO-NOT-REMOVE>
+    #    - name: "feature-by-hostname"
```
done at all places, thanks for the heads up

slintes commented on nfd-worker-daemonset.yaml.template:
```diff
@@ -187,4 +195,31 @@ data:
+#  - name: "more-features-by-hostname-regex"
+#    value: customValue
+#    matchOn:
+#    - hostname:
```
done

slintes commented on nfd-worker-job.yaml.template:
```diff
@@ -197,4 +197,10 @@ data:
+    #        - hostname:
+    #          - worker-0
+    #          - wo.*r-1
+### <NFD-WORKER-CONF-END-DO-NOT-REMOVE>
```
fixed

slintes commented on source/custom/configmap_features.go:
```diff
@@ -0,0 +1,81 @@
+		}
+		log.Printf("DEBUG: custom config rules raw: %s\n", string(bytes))
+
+		config := &config{}
```
fair point, done

ArangoGutierrez commented on nfd-worker-daemonset.yaml.template:
```diff
@@ -187,4 +195,25 @@ data:
+    #        - hostname: ["worker-0", "my-.*-node"]
 ### <NFD-WORKER-CONF-END-DO-NOT-REMOVE>
+---
+## Example for more custom configs in an additional configmap (3/3)
```
Could this be a stand alone file, sameas the `nfd-worker-conf.example` file?

marquiz replied:
Mm, why not. At least, for consistency, this example configmap should be added to the "combined" and "job" templates, too

ArangoGutierrez reviewed (changes requested):
When generating the yamls `make yamls` only the example `nfd-worker-daemonset.yaml.template ` get's populated with the correct mounts and extra ConfigMap , do we want this, or we want all generated examples to carry around empty custom-conf files plus commented lines on how to mount this?

ArangoGutierrez commented on nfd-worker-daemonset.yaml.template:
```diff
@@ -61,6 +61,10 @@ spec:
               readOnly: true
+## Example for more custom configs in an additional configmap (1/3)
+#            - name: more-custom-configs
+#              mountPath: "/etc/kubernetes/node-feature-discovery/custom.d/more-custom-configs"
```
```suggestion
#              mountPath: "/etc/kubernetes/node-feature-discovery/custom.d"
```
just like this, kubernetes will mount the configMap data at `/etc/kubernetes/node-feature-discovery/custom.d/custom.conf`

slintes replied:
oh yes, thanks, in the e2e test I do it correct already :grin:

ArangoGutierrez commented on nfd-worker-daemonset.yaml.template:
```diff
@@ -187,4 +195,25 @@ data:
+#data:
+## Filename doesn't matter, and there can be multiple. They just need to be unique across all additional ConfigMaps
+#  custom.conf: |
+#  - name: "more.kernel.features"
```
just a yaml lint here, line 213 and below need to be moved to the right 2 spaces.
when moved to a single file, you get `error: error parsing custom-configs.yaml: error converting YAML to JSON: yaml: line 9: did not find expected key` meaning everything from line 9 is miss indented

ArangoGutierrez reviewed (changes requested):
Looking good,
```bash
2021/02/16 15:37:18 DEBUG: processing file /etc/kubernetes/node-feature-discovery/custom.d/..data
2021/02/16 15:37:18 DEBUG: skipping hidden file /etc/kubernetes/node-feature-discovery/custom.d/..data
2021/02/16 15:37:18 DEBUG: processing file /etc/kubernetes/node-feature-discovery/custom.d/custom.conf
2021/02/16 15:37:18 DEBUG: custom config rules raw: - name: "more.kernel.features"
  matchOn:
  - loadedKMod: ["example_kmod3"]
- name: "more.features.by.hostname"
  matchOn:
  - hostname: ["ip*"]
 ```
then
```bash
[eduardo@fedora-ws node-feature-discovery-operator]$ oc get no -o json | jq .items[].metadata.labels |grep  custom-more
  "feature.node.kubernetes.io/custom-more.features.by.hostname": "true",
  "feature.node.kubernetes.io/custom-more.features.by.hostname": "true",
```
A small request, could you add some documentation on how to get here, I had to manually create a new configMap, deploy it into the nameSpace I had NFD
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: more-custom-configs
  namespace: openshift-nfd
data:
# Filename doesn't matter, and there can be multiple. They just need to be unique across all additional ConfigMaps
  custom.conf: |
    - name: "more.kernel.features"
      matchOn:
      - loadedKMod: ["example_kmod3"]
    - name: "more.features.by.hostname"
      matchOn:
      - hostname: ["ip*"]
```
then un-comment lines x,y,z from the examples. That should be enough for users trying to leverage this new feature

marquiz commented:
Thanks for the reviews @adrianchiris and @ArangoGutierrez ! I need to call it a day, but, I'll take a look tomorrow

marquiz reviewed (changes requested):
Thanks @slintes for the re-implementation!

Looks really good, overall. Adding the e2e test is really great 👍 

Just a few comments and nits. I also mentioned this in the code comments but I think it'd be better to rename the configMap* terminology to configDir* or smth (you may come up with a better name). This would also mean, renaming `configmap_features.go`

As a side note, there is a lot of debug logging which is basically good, I think. However, our logging is very rudimentary and probably need to address that in a separate PR, adding some control of logger verbosity/filtering.

slintes commented on source/custom/configmap_features.go:
```diff
@@ -0,0 +1,81 @@
+
+// getConfigMapFeatureConfig returns features configured in additional ConfigMaps, which are mounted to
+// /etc/kubernetes/node-feature-discovery/custom.d
+func getConfigMapFeatureConfig() []FeatureSpec {
```
I will go with `getDirectoryFeatureConfig` to be consistent with `getStaticFeatureConfig`

slintes commented on source/custom/configmap_features.go:
```diff
@@ -0,0 +1,81 @@
+		if os.IsNotExist(err) {
+			log.Printf("DEBUG: hostname config directory %v does not exist\n", ConfigMapMountDir)
+		} else {
+			log.Printf("ERROR: unable to access hostname config directory %v\n", ConfigMapMountDir)
```
:+1:  (and I need to get used to use %q for strings, lazy me almost always uses %v for everything :grin: )

slintes commented on source/custom/rules/hostname.go:
```diff
@@ -0,0 +1,54 @@
+
+		match, err := regexp.MatchString(hostnamePattern, nodeName)
+		if err != nil {
+			log.Printf("ERROR: error testing regex: pattern %s, nodename %s, %v", hostnamePattern, nodeName, err)
```
makes sense :+1:

slintes commented:
Thanks for the reviews and feedback all :+1: 
Most things are addressed:
- Renamed hostname -> nodename at all places
- Use more generic name for the additional configs (ConfigMap -> Directory)
- Improved and less logging (I commented some debug logs for reducing amount of logs, can be re-enabled when switching to a logging framework with configurable verbosity... but that's out of scope of this PR :slightly_smiling_face: )
- and some smaller things

About the manifests: should I add the ConfigMap and its mounts to all manifests? I'd prefer to mention the daemonset in documentation as example instead of copying more yaml snippets around :thinking:  

Talking about doc: that's the next item on my list...

slintes commented:
Updated PR title and description to reflect recent changes.
And updated doc :slightly_smiling_face:

ArangoGutierrez commented:
/label needs-rebase

ArangoGutierrez commented:
meh, ok @slintes a simple rebase will fix the CI after your PR update

slintes commented:
rebased, and squashed commits

ArangoGutierrez commented:
I think during the squash event you removed the tests

marquiz commented on docs/get-started/features.md:
```diff
@@ -124,15 +124,25 @@ See the full list in [github.com/klauspost/cpuid][klauspost-cpuid].
+
+#### Additional configuration directory
+
+Additionally to the rules defined in the nfd-worker configuration file, the Custom feature can read more configuration
+files located in the `/etc/kubernetes/node-feature-discovery/custom.d/` directory. This makes more dynamic and flexible
+configuration easier. This directory must be available inside the NFD worker container, so Volumes and VolumeMounts must
+be used for mounting e.g. ConfigMap(s). The
+[nfd-worker deployment template](https://github.com/kubernetes-sigs/node-feature-discovery/blob/{{ site.release
+}}/nfd-worker-daemonset.yaml.template#L204)
+provides an example for providing Custom configuration with an additional ConfigMap, mounted into the `custom.d`
+directory.
```
nit: could you wrap the lines to 80 chars (also elsewhere in this file)? Markdown linter (`mdl`) bugs about these. I haven't integrated that to CI because we don't fully pass muster

slintes replied:
sure, thx, done

marquiz commented on nfd-worker-daemonset.yaml.template:
```diff
@@ -68,6 +68,10 @@ spec:
             - name: nfd-worker-conf
               mountPath: "/etc/kubernetes/node-feature-discovery"
               readOnly: true
+## Example for more custom configs in an additional configmap (1/3)
+#            - name: more-custom-configs
+#              mountPath: "/etc/kubernetes/node-feature-discovery/custom.d/more-custom-configs"
+#              readOnly: true
```
For consistency, add this to the "combined" and "job" templates, too

marquiz commented on nfd-worker-daemonset.yaml.template:
```diff
@@ -94,6 +98,10 @@ spec:
         - name: nfd-worker-conf
           configMap:
             name: nfd-worker-conf
+## Example for more custom configs in an additional configmap (2/3)
+#        - name: more-custom-configs
+#          configMap:
+#            name: more-custom-configs
```
For consistency, add this to the "combined" and "job" templates, too

slintes replied:
done

marquiz commented on nfd-worker-daemonset.yaml.template:
```diff
@@ -199,4 +207,25 @@ data:
+#apiVersion: v1
+#kind: ConfigMap
+#metadata:
+#  name: more-custom-configs
```
nit: maybe it's my problem, but this is just somehow funny name 😄 . How about e.g. `custom-source-extra-config`, `custom-source-extra-rules` , `custom-extra-rule` or smth (you can probably come up with better name)?

marquiz reviewed (changes requested):
@slintes thank you for the refresh. Few small comments, nothing big.

One thing I'm a bit dubious are the now-commented-out log messages. But, I think that in the light of the much-needed (and upcoming) logging rework we can leave them there (and uncomment with the logger patch)

slintes commented:
Thanks for review, all remarks are addressed.

> One thing I'm a bit dubious are the now-commented-out log messages. But, I think that in the light of the much-needed (and upcoming) logging rework we can leave them there (and uncomment with the logger patch)

yes, that's the idea

adrianchiris commented on docs/get-started/features.md:
```diff
@@ -328,6 +361,10 @@ custom:
+  - name: "my.datacenter"
+    value: "datacenter-1"
+    matchOn:
+      - hostname: [ "node-datacenter1-rack.*-server.*" ]
```
nodename ?

slintes replied:
good catch, thanks!

adrianchiris commented on docs/get-started/features.md:
````diff
@@ -291,6 +304,26 @@ Matching is done by performing logical _AND_ for each provided Element, i.e the
+Element     :A nodename regexp pattern
+```
+
+The Rule allows matching the node's name against a provided list of Elements.
````
Should a note be added here clarifying that this matches on the node's name as registered in the kubernetes cluster ?
or do you think its clear they way it is now ?

slintes replied:
IMHO node name is clear...?

adrianchiris replied:
ack

adrianchiris commented on nfd-daemonset-combined.yaml.template:
```diff
@@ -232,4 +240,25 @@ data:
+#    - name: "more.features.by.nodename"
+#      value: customValue
+#      matchOn:
+#      - nodename: ["special-.*-node-.*"]
```
nit: missing newline

slintes replied:
fixed

adrianchiris commented on nfd-worker-job.yaml.template:
```diff
@@ -204,4 +212,25 @@ data:
+#    - name: "more.features.by.nodename"
+#      value: customValue
+#      matchOn:
+#      - nodename: ["special-.*-node-.*"]
```
nit: missing newline

adrianchiris commented on source/custom/custom.go:
```diff
@@ -90,58 +98,35 @@ func (s Source) Discover() (source.Features, error) {
-			if !match {
-				continue
+			} else if !match {
+				ruleFailed = true
```
not sure if its favored but you could use `continue <label>` statement to continue to the next `matchRule` on the outer loop.

You would not need the `ruleFailed` bits in this case.
(just a suggestion here)

adrianchiris replied:
yea no worries, im fine either way.

marquiz replied:
> Personally I dislike those labels...

I'm not a big fan either, even if I've used them one or two times. You could achieve the above without labels e.g. by moving the inner loop into a separate function. But I think it's fine the way it is now and leave this to you to decide

slintes replied:
I like the idea of moving the inner loop into a func, will do that, thanks

adrianchiris commented:
commit message contains some typos, can be fixed when merged or if an update is pushed to this PR

slintes commented on docs/get-started/features.md:
```diff
@@ -134,6 +134,18 @@ examples how to set-up and manage the worker configuration.
+dynamic and flexible configuration easier. This directory must be available
+inside the NFD worker container, so Volumes and VolumeMounts must be used for
+mounting e.g. ConfigMap(s). The
+[nfd-worker deployment template](https://github.com/kubernetes-sigs/node-feature-discovery/blob/{{ site.release }}/nfd-worker-daemonset.yaml.template#L204)
```
Fair point. But I removed the link completely, because the example config is in all manifests now.

slintes commented:
> Overall looks OK, some minor nits.

thanks! Pushed another new version :slightly_smiling_face:

adrianchiris commented on docs/get-started/features.md:
```diff
@@ -134,6 +134,17 @@ examples how to set-up and manage the worker configuration.
+#### Additional configuration directory
+
+Additionally to the rules defined in the nfd-worker configuration file, the
+Custom feature can read morddde configuration files located in the
```
>  morddde
typo? must have missed it last round of review.

slintes replied:
You didn't miss it, I messed it up somehow during reformat in the latest push. Thx and Fixed.

marquiz commented on source/custom/directory_features.go:
```diff
@@ -0,0 +1,81 @@
+	features := make([]FeatureSpec, 0)
+
+	log.Printf("DEBUG: getting files in %s", Directory)
+	files, err := ioutil.ReadDir(Directory)
+	if err != nil {
+		if os.IsNotExist(err) {
+			log.Printf("DEBUG: custom config directory %q does not exist", Directory)
+		} else {
+			log.Printf("ERROR: unable to access custom config directory %q, %v", Directory, err)
+		}
+		return features
+	}
+
+	for _, file := range files {
+		configFile := filepath.Join(Directory, file.Name())
+
+		if file.IsDir() {
+			//log.Printf("DEBUG: skipping dir %q", configFile)
+			continue
```
Hmm, I started to think that maybe we should recurse into subdirectories as well. At least one level deep. This would make it much more user-friendly to drop in (mount) multiple config files under `custom.d/`. Each configmap would just be mounted in a separate subdir, that is.

I think with the current scheme it is only possible to use multiple configmaps with `subPath` mount which a bit cumbersome (IMO), and more importantly, prevents nfd-worker from getting confimap updates (i.e. the config inside the container is static even if the configmap(s) changed).

slintes replied:
Oh, good catch! I thought I can mount multiple CMs into the same path as long as the filenames in their data is unique, but that doesn't work. Will fix this and extend the e2e test to use 2 CMs.

marquiz replied:
> I thought I can mount multiple CMs into the same path as long as the filenames in their data is unique, but that doesn't work.

Unfortunately that's not the case 😞  It's possible with the `subPath` directive, but, it has it's drawbacks as explained above.

> Will fix this and extend the e2e test to use 2 CMs.

That would be great 👍

slintes commented:
- refactored the nested loops in custom.go
- read config files from 1st level subdirs of custom.d and test multiple CMs in e2e test

For easier review commits are not squashed yet, happy to do so after review

marquiz reviewed (changes requested):
Nice work @slintes, and really fast response times(!) 😄  I came up with two more comments. We're really close, I think...

slintes commented on nfd-daemonset-combined.yaml.template:
```diff
@@ -115,6 +115,11 @@ spec:
+## When using multiple configmaps, mount them into subdirectories of custom.d
+#            - name: custom-source-extra-rules
+#              mountPath: "/etc/kubernetes/node-feature-discovery/custom.d/"
+#              readOnly: true
```
works for me, done

slintes commented on test/e2e/node_feature_discovery.go:
```diff
@@ -581,23 +590,183 @@ var _ = framework.KubeDescribe("[NFD] Node Feature Discovery", func() {
+				targetNodeName := ""
+				for _, node := range nodeList.Items {
+					for k := range node.Labels {
+						if k == "node-role.kubernetes.io/worker" {
```
good point, thanks and done

marquiz commented on test/e2e/node_feature_discovery.go:
```diff
@@ -626,7 +626,7 @@ var _ = framework.KubeDescribe("[NFD] Node Feature Discovery", func() {
 				for _, node := range nodeList.Items {
 					for k := range node.Labels {
-						if k == "node-role.kubernetes.io/worker" {
+						if k != "node-role.kubernetes.io/master" {
```
There are many labels which are not `node-role.kubernetes.io/master` 😉 

Suggest to do smth like this
```
for _, node := range nodeList.Items {
	if _, ok := node.Labels["node-role.kubernetes.io/master"]; !ok {
		targetNodeName = node.Name
		break
	}
}
```

slintes replied:
LOL, sorry... I should wait for finishing the 1st coffee in the morning before doing such "trivial changes" :D

slintes replied:
fixed, thx

marquiz commented:
Looks good and I don't think I've anything to add. I think you could squash the commits, now

slintes commented:
> Looks good and I don't think I've anything to add. I think you could squash the commits, now

thanks & done :)

marquiz reviewed (changes requested):
Ach, one more nit, about the commit message.

I've been quite lazy in reviewing the commit messages, probably because that's somewhat inconvenient in Github. I think it would be good practice for us to try to follow the k8s [commit message guidelines](https://www.kubernetes.dev/docs/guide/pull-requests/#commit-message-guidelines).

I'd like to address at least *Use imperative mood in your commit message subject* and *Providing additional context* by changing the subject to something like:
`source/custom: add nodename rule`

Also the commit message body could be a bit more better read but I think I can live with that. Even if it's quite draft-like and I'm not a big fan of a bullet list only.

Another note: there are some trailing whitespace in `test/e2e/node_feature_discovery.go`

marquiz commented:
> Also the commit message body could be a bit more better read but I think I can live with that. Even if it's quite draft-like and I'm not a big fan of a bullet list only.

Talking about the commit message body, you already addressed *[Use the commit message body to explain the what and why of the commit](https://www.kubernetes.dev/docs/guide/pull-requests/#use-the-commit-message-body-to-explain-the-_what_-and-_why_-of-the-commit)* in the [description of this PR](https://github.com/kubernetes-sigs/node-feature-discovery/pull/429#issue-566745454). Some of that could be re-used in the commit message, I think.

marquiz commented on test/e2e/node_feature_discovery.go:
```diff
@@ -581,23 +590,181 @@ var _ = framework.KubeDescribe("[NFD] Node Feature Discovery", func() {
+
+				data2 := make(map[string]string)
+				data2["custom1.conf"] = `
+- name: ` + targetLabelNameWildcard + `  
```
trailing whitespace

marquiz commented on test/e2e/node_feature_discovery.go:
```diff
@@ -581,23 +590,181 @@ var _ = framework.KubeDescribe("[NFD] Node Feature Discovery", func() {
+  matchOn:
+  - nodename:
+    - ` + targetNodeNameWildcard + `
+- name: ` + targetLabelNameNegative + `  
```
trailing whitespace

slintes commented:
fixed whitespaces and improved commit message

ArangoGutierrez commented:
```bash
Ran 3 of 220 Specs in 62.471 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 217 Skipped
--- PASS: TestE2E (62.47s)
PASS
ok      sigs.k8s.io/node-feature-discovery/test/e2e     62.575s
```

ArangoGutierrez commented:
/retest all

ArangoGutierrez commented:
/retest

ArangoGutierrez commented:
/test all

ArangoGutierrez commented on source/custom/rules/nodename_rule.go:
```diff
@@ -0,0 +1,51 @@
+			continue
+		}
+		if !match {
+			log.Printf("DEBUG: nodename rule: No match for for pattern %q with node %q", nodenamePattern, nodeName)
```
Double `for for` in log string

slintes replied:
fixed... and in the light of "too many log messages" I commented it and the other one out

ArangoGutierrez reviewed (changes requested):
One last catch

ArangoGutierrez commented on source/custom/rules/nodename_rule.go:
```diff
@@ -0,0 +1,50 @@
+			//log.Printf("DEBUG: nodename rule: No match for pattern %q with node %q", nodenamePattern, nodeName)
+			continue
+		}
+		//log.Printf("DEBUG: nodename rule: Match for pattern %q with node %q", nodenamePattern, nodeName)
```
@marquiz you think that after the #445 this 2 log's could be DEBUG lvl, only seen when requested?

marquiz replied:
Yeah, I think so

marquiz reviewed (approved):
Thank you @slintes for your efforts and patience! 😄  This is a great enhancement to NFD
/approve

ArangoGutierrez reviewed (approved):
Awesome, can't wait to support this via the Operator as well 
/lgtm

slintes commented:
Thanks all for the very helpful and constructive reviews :+1:
